distributed/worker.py

```diff
@@ -3408,34 +3416,44 @@ async def find_missing(self) -> None:
 
-    def update_who_has(self, who_has: dict[str, Collection[str]]) -> None:
-        try:
-            for dep, workers in who_has.items():
-                if not workers:
-                    continue
+    def _update_who_has(self, who_has: Mapping[str, Collection[str]]) -> None:
+        for key, workers in who_has.items():
+            ts = self.tasks.get(key)
+            if not ts:
+                # The worker sent a refresh-who-has request to the scheduler but, by the
+                # time the answer comes back, some of the keys have been forgotten.
+                continue
+            workers = set(workers)
+
+            if self.address in workers:
+                workers.remove(self.address)
+                if ts.state != "memory":
+                    # This can happen if the worker recently released a key, but the
+                    # scheduler hasn't processed the release-worker-data message yet.
+                    # It is not necessary to send a missing-data message back to the
+                    # scheduler here, because it is guaranteed to reach the scheduler
+                    # after the release-worker-data from the same BatchedSend channel.
+                    logger.debug(
+                        "Scheduler claims worker %s holds data for task %s, "
+                        "which is not true.",
+                        self.address,
+                        ts,
+                    )
 
-                if dep in self.tasks:
-                    dep_ts = self.tasks[dep]
-                    if self.address in workers and self.tasks[dep].state != "memory":
-                        logger.debug(
-                            "Scheduler claims worker %s holds data for task %s which is not true.",
-                            self.name,
-                            dep,
-                        )
-                        # Do not mutate the input dict. That's rude
-                        workers = set(workers) - {self.address}
-                    dep_ts.who_has.update(workers)
+            if ts.who_has == workers:
+                continue
 
-                    for worker in workers:
-                        self.has_what[worker].add(dep)
-                        self.data_needed_per_worker[worker].push(dep_ts)
-        except Exception as e:  # pragma: no cover
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+            for worker in ts.who_has - workers:
+                self.has_what[worker].discard(key)
+                # Can't remove from self.data_needed_per_worker; there is logic
+                # in _select_keys_for_gather to deal with this
 
-                pdb.set_trace()
-            raise
+            for worker in workers - ts.who_has:
+                self.has_what[worker].add(key)
+                if ts.state == "fetch":
+                    self.data_needed_per_worker[worker].push(ts)
```

Comment:
Why the change? There's no reason to push it into the heap if state=memory or flight.
See 
https://github.com/dask/distributed/blob/5e9e97fe39b341061849e370cd386ca1a4447d17/distributed/worker.py#L2005-L2014

which is a helper function exclusively of
- transition_missing_fetch
- transition_released_fetch
- transition_flight_fetch

(the other two transition_\*_fetch functions don't call it because they don't set the state to fetch).

Reply:
Yes, as I said, it feels odd but I understand. Let's move on